nn.py

```diff
@@ -412,51 +417,52 @@ class SubModel(nn.Module):
     Attributes:
         lstm       = [SplitLSTM] LSTM part of the submodel
         fc         = [SplitLinear] fully-connected part of the submodel
-        num_groups = [int] number of groups this submodel will process
     """
 
-    def __init__(self, num_out, num_groups, independent):
+    def __init__(self, num_groups, independent):
         """Initializes the submodel.
 
         Args:
-            num_out     = [int] number of output units per store-item group
-            num_groups  = [int] number of store-item groups to make submodel for
             independent = [bool] whether the submodel has independent groups
         """
         super(SubModel, self).__init__()
 
-        self.num_groups = num_groups
-        self.lstm = SplitLSTM(self.num_groups, independent)
-        self.fc = SplitLinear(num_out, self.num_groups, independent)
+        self.lstm = SplitLSTM(num_groups, independent)
+        self.fc = SplitLinear(num_groups, independent)
 
     def reset_hidden(self):
         """Resets the hidden state of the LSTM."""
         self.lstm.reset_hidden()
 
-    def forward(self, day, items):
+    def forward(self, day, items, t_day, t_items):
         """Forward pass of the submodel.
 
         Args:
             day   = [torch.Tensor] inputs constant per store-item group
                 The shape should be (1, seq_len, num_const).
             items = [torch.Tensor] inputs different per store-item group
                 The shape should be (1, seq_len, num_groups, num_var).
+            t_day   = [torch.Tensor] targets unequal per store-item group
+                The shape should be (1, horizon, num_const).
+            t_items = [torch.Tensor] targets unequal per store-item group
+                The shape should be (1, horizon, num_groups, num_var).
```

Comment:
What are these targets here (in comparison to the inputs)? I see that you use them below to run the LSTM again, but I don't quite understand what this does.

Reply:
Let's try to explain it with an example.

Let's set seq_len = 8 and horizon = 5 and let's say that the start index is equal to 3.

Then the first iteration, data from index 3 up to and including 10 is returned as inputs
and data from index 11 up to and including 15 is returned as targets.
The input data is fed through the LSTM, which saves its hidden state at index 10.
Then the target data is fed through the LSTM, but it keeps the hidden state at index 10.
The output from the second pass through the LSTM is turned into a batch with horizon (5) elements and each of the batch elements is fed through the FC layer.
The output is then (horizon, num_groups) or (5, 30490) after concatenating the output from each submodel.

The second iteration, the hidden state of the LSTM is still at index 10, the input data is from index 11 up to and including 18 and the targets data is from index 19 up to and including 23. After running all the data through the LSTM and FC layer, the hidden state is now set at index 18 and the output is again (5, 30490).

This goes on until the end of an epoch.

If seq_len = 1, then the input data has 1 element, but the targets data still has 5 elements. So, each iteration, the index of the hidden state of the LSTM is increased by 1, while the size of the output stays (5, 30490). This is a sliding window, because we advance the index by 1 each iteration, while having an output batch with more than one element.

If horizon = 1, then each iteration the input data has 8 elements, the targets data has 1 element, and the output has size (1, 30490). This is not the same as with the old code, where we had (1, 5, 30490) as output. But you do advance the index of the hidden state of the LSTM with `seq_len` each iteration, where the number of elements in a batch is equal to 1. So you skip some indices, where you do not test the performance on all days each epoch, which is sort of what we were doing with the old code.

Please ask questions if this wasn't clear!

Reply:
The target data also have the target sales used to compute the loss from the second column onwards. These are very bad names (feel free to come up with better ones), but I have to split the data into inputs and 'targets' to be able to save the intermediate hidden state of the LSTM.

Reply:
Thanks for the elaborate answer! It is clear to me now I think. So `t_day` and `t_items` are the features for the target sales, but we don't use the target sales of course right (because that is what we are predicting)?

I understand the `seq_len` and `horizon` part now with the example you provided 👍.

Reply:
Yes. So the size of the targets is actually (`horizon + 1`, *). We use all the features of the first up to and including the `horizon`th element to get the projected sales. Then we compute the loss with the outputs of the network and the sales of the second up to and including the `horizon + 1`th element.